drivers/network/dd/e1000/hardware.c

```diff
@@ -189,32 +190,75 @@ static BOOLEAN E1000ReadMdic(IN PE1000_ADAPTER Adapter, IN ULONG Address, USHORT
 }
 #endif
 
-
 static BOOLEAN E1000ReadEeprom(IN PE1000_ADAPTER Adapter, IN UCHAR Address, USHORT *Result)
 {
     ULONG Value;
     UINT n;
 
-    E1000WriteUlong(Adapter, E1000_REG_EERD, E1000_EERD_START | ((UINT)Address << E1000_EERD_ADDR_SHIFT));
+    if (Adapter->EERDAddrShift == 0xffff)
+    {
+        NDIS_DbgPrint(MIN_TRACE, ("E1000DetermineAndSaveEERDLayout not run or wasn't successfull!\n"));
+        return FALSE;
+    }
+    E1000WriteUlong(Adapter, E1000_REG_EERD, E1000_EERD_START | ((UINT)Address << Adapter->EERDAddrShift));
 
     for (n = 0; n < MAX_EEPROM_READ_ATTEMPTS; ++n)
     {
         NdisStallExecution(5);
 
         E1000ReadUlong(Adapter, E1000_REG_EERD, &Value);
 
-        if (Value & E1000_EERD_DONE)
+        if (Value & Adapter->EERDDoneBit)
             break;
     }
-    if (!(Value & E1000_EERD_DONE))
+    if (!(Value & Adapter->EERDDoneBit))
     {
-        NDIS_DbgPrint(MIN_TRACE, ("EEPROM Read incomplete\n"));
+        NDIS_DbgPrint(MIN_TRACE, ("EEPROM Read incomplete! %x \n", Value));
         return FALSE;
     }
     *Result = (USHORT)(Value >> E1000_EERD_DATA_SHIFT);
+    NDIS_DbgPrint(MIN_TRACE, ("Read: %02x %04x\n", Address, *Result));
     return TRUE;
 }
 
+/* 8254x and 8257x have different layouts of EERF
+   Get it by trying to read the VID-field.
+*/
+static BOOLEAN E1000DetermineAndSaveEERDLayout(IN PE1000_ADAPTER Adapter) {
+    BOOLEAN Success;
+    USHORT Result = 0;
+
+    /* First, try the 8254x layout */
+    Adapter->EERDDoneBit = E1000_EERD_DONE_8254x;
+    Adapter->EERDAddrShift = E1000_EERD_ADDR_SHIFT_8254x;
+
+    Success = E1000ReadEeprom(Adapter, E1000_NVM_REG_VID, &Result);
+    if (Success && Result == NVM_VID) {
```

Comment:
Code style (for your next modifications): curly brackets on their own lines.

Reply:
Done